### ISSUE TITLE
Add a Version String

### ISSUE THREAD

**Issue by viluon** [zdevito/terra#242]
Unlike Lua (5.*) and LuaJIT, Terra has no `-v` command-line option. It also lacks a global variable containing the version string. This would be very useful to have for automated builds, for example.

It's a pity that the project doesn't follow [semantic versioning](semver.org). The release tag or build options would be equally handy for determining which version of Terra is currently in use, which would in turn greatly ease any debugging processes.

**Comment by Qix-** [comment]
Semantic versioning isn't the end-all-be-all. There are flaws with semver - read about some of them [here](https://gist.github.com/jashkenas/cbd2b088e20279ae2c8e) - and given that compressing API changes between Lua/C/C++/dependency versions (LLVM, Lua, etc.) as well as any functionality/behavior changes into a single number is *impossible*, semantic versioning wouldn't help anyone other than to add extra stress and arguing into the versioning process.

So +1 on the version string/flags, -1 on semver.

**Comment by OvermindDL1** [comment]
Semantic versioning is perfectly suited for versioning API's (though calver should be used for application releases that are not API dependent).  That gist is quite full of quite a variety of logical fallacies.

**Comment by viluon** [comment]
I second @OvermindDL1. Read a couple of the longer comments, what the author complains about is pretty much unrelated to the version format.

I've yet to see a version format which compresses functionality, behaviour, and dependencies changes into a single number. SemVer does none of that and still uses at least three numbers. Please provide an example of such a useful concept, @Qix-.

**Comment by Qix-** [comment]
@viluon I can't really give you concrete examples except for the fact that I manage [quite a few](https://npmjs.org/~qix) Node.js packages and I run into issues and arguments about semver _all the time_ regarding various things pretty neatly laid out in the article I linked.

I'd be interested in why you think the article doesn't address anything about the version format.

> Ultimately, SemVer is a false promise that appeals to many developers — the promise of pain-free, don't-have-to-think-about-it, updates to dependencies. But it simply isn't true.

Actually _following_ semver is painful and generally impossible. Representing complex, nuanced changes such as "a dependency changed that might break _some_ users but not all" or "a very non-critical API endpoint was removed" (instead of a _very_ critical API endpoint) are not representable using semver - but both require a major release if you strictly follow semver.

Also, what is patch and what is minor is always a cause for concern. Plus, pretty much no package manager for native code utilizes semver, and since semver was created for robots (specifically, [npm](https://github.com/npm/npm)) that doesn't really leave native code with a lot to gain from it.

Semver would work if humans were perfect. We aren't. I have a **lot** of experience with working with semver and releasing production packages with it. It's not worth the trouble.

**Comment by OvermindDL1** [comment]
> Actually following semver is painful and generally impossible. Representing complex, nuanced changes such as "a dependency changed that might break some users but not all" or "a very non-critical API endpoint was removed"

Honestly if this is what you are trying to do then you are using semver wrong, the semver website very clearly states when each number should be bumped or reset, and such things like if an API changed in such a way that it might even break backwards compatibility then you bump the major version, no ifs-and-or-buts.

SemVer is purely for API versioning, and yes some people do use it when a calver should be used, but API's absolutely should be versioned via SemVer.

> So +1 to a version string in the terra CLI, -1 very strongly to semver.

Except a simple version string would not state when I need to look over my code.  A patch or less version change means I can and should update with impunity.  A minor version change means I should update with impunity *and* I should also see if there are new features I can find useful.  A major update I should not touch until I give a thorough look over the changelog and API.

Where there is generally only a single calver for an entire project, different 'contexts' of an API would also have their own semver's, with the overall API semver being an aggregation there-of if it is necessary at all to have one (a single include to bring in all the others for example).

What you prior article was speaking of was primarily of people mis-using it, nothing about semver being poor itself, and there is nothing listing what could be better (feature flags are utterly horrid for that as that then means you have a *lot* of things to check and go over on every-single-update as one big issue).

I don't know about the node community, I try to stay away from poor languages and ecosystems as such, but in the C++, Rust, Beam, and OCaml worlds it has been rock solid, but it also tends to be very strictly followed.

Simple 'countable progress' as you put it tells nothing about feature addition or removal or how safe it is to upgrade to a given version, it's nothing but an even more simple CalVer (except unlike CalVer it wouldn't even imply how old it is).

**Comment by Qix-** [comment]
> What you prior article was speaking of was primarily of people mis-using it, nothing about semver being poor itself, and there is nothing listing what could be better (feature flags are utterly horrid for that as that then means you have a lot of things to check and go over on every-single-update as one big issue).

Exactly - which is what happens all the time in the Node.js community. And if you really want to argue that all Node.js developers are misusing semver whereas the rest of the world uses it perfectly, by all means make that assertion. The node.js community made semver popular.

Nothing is wrong with semver at the theoretical level (except that I'd prefer [explicit versioning](https://medium.com/sapioit/why-having-3-numbers-in-the-version-name-is-bad-92fc1f6bc73c) over semver) but that's not what's being argued. It's how people *use* it that makes it useful or not useful.

> Simple 'countable progress' as you put it tells nothing about feature addition or removal or how safe it is to upgrade to a given version, it's nothing but an even more simple CalVer (except unlike CalVer it wouldn't even imply how old it is)

What? Semver has countable progress. I didn't say a literal counter. However, I will say that if you're relying on a single version number to tell you if a dependency should be upgraded and how suitable it is for your use-case, you're in for a world of hurt.

Further, the decision to upgrade a dependency can't just be based on if its API has changed. That's preposterous.

**Comment by viluon** [comment]
@Qix-, what even is your point here? I opened this issue because I lacked the version information for Terra in automated builds. I could easily use calver or some other poorly standardized versioning scheme, _if it were added_. Clearly, _some_ distinction between Terra releases is still better than _none_. But if a version string were to be added, it should be of the SemVer format.

I couldn't care less about your npm issues. SemVer is the **most widely used, standardized, and sane version format**. I don't care who made it popular.

The SemVer spec is well-defined, to-the-point, clear. You (and the gist you linked) are whining about shady features SemVer _never meant to cover_. I failed to find a mention of the "false promise" you're talking about in the SemVer spec. It **does not guarantee** "pain-free, don't-have-to-think-about-it, updates to dependencies," and it never did. What SemVer tries to accomplish is avoiding [dependency hell](https://en.wikipedia.org/wiki/Dependency_hell). Judging by its widespread usage ([1](https://luarocks.org/), [2](https://crates.io/), [3](https://rubygems.org/), [4](https://semver.io/), [5](https://blog.npmjs.org/post/162134793605/why-use-semver), many many others), I'd say it does a pretty good job at that. Most of my Arch Linux packages follow semantic versioning, as does the Linux kernel, and I have yet to run into a problem caused by the version format.

You said it yourself
> It's how people use it that makes it useful or not useful.

Stop shooting a good standard down because of people misusing it. That's ridiculous. There is no reason to believe that Terra would misuse SemVer too.

**TL;DR**: Read the [semver spec](http://semver.org). The rules work very well if followed carefully. The format is straightforward.

**Comment by Qix-** [comment]
> @Qix-, what even is your point here? 

> But if a version string were to be added, it should be of the SemVer format.

Firstly, you seem heated. We're attempting to come to a solid solution here and I'm dissenting from the proposal of using semver for versioning. I've said twice now that a version string is fine. The format is what I'm calling into question.

> SemVer is the most widely used, standardized, and sane version format.

Citation needed. I've not hardly seen any C packages use Semver for the direct reason that "API" means a *lot* of things since C is not as straightfoward to describe in just a few short numbers.

API in terms of terra could mean any one of

- The Lua API
- The `terralib` API
- The API exposed by FFI
- The API exposed by LuaJIT
- The CLI options
- The ABI used (calling conventions, etc.)
- Terra's use of the Clang frontend (and how it lays out members)
- The Pratt parser API that is exposed
- The Terra syntax superset
- The Lua syntax subset (e.g. if Lua were to be updated)

I'm sure there are others as well.

All of the above affect downstream users. Semver cannot begin to adequately describe all of the nuanced changes there. Further, when you start getting into Semver ranges, you see even more problems - all stemming from misuse of Semver. Whether you like it or not, Semver bumps (and the changes therein) cannot be verified by machines and thus the standard is prone to human error (and misuse).

Calver doesn't solve your problem either, I'm not entirely sure why it keeps getting thrown around (calver doesn't really add much _useful_ information that a numeric, incremental counter would provide anyway - both of which leave much to be desired).

> What SemVer tries to accomplish is avoiding dependency hell.

If you're referring to semver ranges, would you rather pinned version numbers or demonic Github bots such as @greenkeeperio that spam the daylights out of your repositories with PRs and issues regarding outdated dependencies?

Dependency hell happens when you have nested dependencies relying on different versions of the same upstream dependency (e.g. **A** depends on **B v0.1** and **C**, and **C** depends on **B 0.4**). This doesn't really happen in the case of C since most of the time those situations cause duplicate symbol errors anyway.

That argument is weak at best - the ecosystem Terra uses is entirely different.

> as does the Linux kernel

Not officially. They have a three-letter versioning system when they choose to but do not subscribe to Semver (Torvalds has been using a similar version format since long before Semver 1 was written).

> I have yet to run into a problem caused by the version format.

That just means you haven't worked with it enough ;) Don't discount my concerns because of (what I assume is) your lack of experience with shipping lots of production Semver-based code. The argument is anecdotal, as is mine - mine just happens to be fueled by many hours of headaches and anguish.

> Stop shooting a good standard down because of people misusing it. That's ridiculous.

Your assertion that it's a "good standard" is a fallacy here - if it were a good standard, there would be no question about it and the arguments against it would be weak or non-existent. That clearly isn't the case.

What's ridiculous is the amount of anger resulting from a dissenting opinion. Choose Semver or not, I wanted to make a point to disagree with the motion to use Semver. That's all.

> There is no reason to believe that Terra would misuse SemVer too.

Humans are humans. You simply cannot verify the accuracy of a Semver version bump programmatically, or else we wouldn't really need Semver at all (as there would be a much better versioning scheme that took humans out of the equation entirely). Therefore, it's very much prone to error given the laundry list of things in this project alone that one could consider an "API".

If Terra was merely a small static/shared library with a straightforward C API and some headers, this would be a different discussion.

> TL;DR: Read the semver spec. The rules work very well if followed carefully. The format is straightforward.

I've read it a good hundred times. I've implemented Semver utilities in a few languages, quite a few times. It's not perfect. It's not even great if you ask me.

**Comment by OvermindDL1** [comment]
Too long to respond to in depth in the short period that I have, but you really need to lay off the ad hominem attacks as it makes your entire argument look invalid regardless of any merit it has.

However, from https://semver.org/ this:

> Additional labels for pre-release and build metadata are available as extensions to the MAJOR.MINOR.PATCH format.

And it details how those should appear and be formatted.  You seem to keep stating that semver is purely 3-integer when it is not.

As for the terra has too many parts to version, that is why there are multiple versions for each different subsystem with the overall semver being bumped as appropriate any time any of those bump (which generally means that it tends to have a significantly higher major version), this is what I stated before as well yet it was ignored while still focusing on the strawman.  You are asking for a countably increasing version scheme, semver gives you that and more, the 'more' you can ignore for your own uses if you wish but it is highly valuable to those that consume it as it gives an easy overview to tell whether the changelog/documention needs to be looked at for a given upgrade.

**Comment by viluon** [comment]
Sorry @Qix-, you're right -- I was heated. I don't know how else to explain this to you. Time and time again we show you that your understanding of SemVer is wrong. What you think it fails at is not what it was meant to do. You complain about ranges and package manager issues, which the format is not responsible for. Ranges and issues like that happen with any version format. Your apparent disgust at SemVer is irrational. And you have failed to suggest an alternative format which wouldn't suffer of these issues.

> C is not as straightfoward to describe in just a few short numbers.

SemVer isn't a ... whatever you're suggesting it is. It's not a means of describing programming languages, that's for sure.

> "API" in terms of terra could mean any one of
> - The Terra Lua (C) API (terra_loadfile(), etc)
> - The terralib API
> - The API exposed by FFI
> - ...

No. No no no no no. You got it all wrong! That's not how versioning works at all. _Your_ version number should only say what SemVer says it should - nothing more! I mean it when I say you should read the spec again. Version number doesn't say what dependencies changed. It doesn't list their versions. It only says that a change was introduced, and it was one of three kinds:
1. a major change, possibly breaking things
2. a minor change, possibly adding things
3. just a patch fixing broken features

What you seem to demand from SemVer is insane -- it's not supposed to describe everything you listed at all!

A version number is a **singular** thing. If Lua is updated say from 5.1.3 to 5.1.4 in a new Terra release, follow the rules! (Let's pretend Terra is using SemVer and its current release is 1.0.0)
1. Is it a breaking change? -- No, in Lua, it's just a patch! (note that Lua does not follow SemVer, a bump in the middle number would mean breaking changes)
2. Is it a feature addition? -- No! No new features were added
3. Is it a fix? -- Yes! Let's **bump Terra to 1.0.1**

There! Now what if the C API changed? Let's say a new function got added, `terra_foo`. Now then, what to do with our version (v1.0.1)?
1. Is it a breaking change? -- No
2. Is it a feature addition? -- Yes! We can stop here. A new feature was added!

Well then, that's a minor version bump! Let us change the version number appropriately: `v1.0.1 --> v1.1.0`

Finally, let's pretend that `terralib` got a new function, `bar()`, and the previously included `terra_foo` was removed. The addition of `bar()` in itself would mean a minor version bump, but the removal cannot be ignored - let's see what the rules say:
1. Is it a breaking change? -- Yes! Some code may depend on `terra_foo`, so we must do the sane thing and bump the major version number.

That's right, with just a simple removal a major version bump is required. And that's all right! We must inform all our users of the breaking change, no matter how tiny it may look from our point of view. Terra goes from `v1.1.0` to `v2.0.0`! Needless to say, breaking changes should be avoided as much as possible.

"But what about the other things?" You may ask, wondering why did I ignore things like "Terra's use of the Clang frontend" from your list. Maybe you didn't read the [spec](https://semver.org/) carefully enough. Let me remind you:
> 1. Software using Semantic Versioning MUST declare a public API. This API could be declared in the code itself or exist strictly in documentation. However it is done, it should be precise and comprehensive.

The public API of Terra is the language itself (Terra + Lua) and its standard library. Is the way it uses a dependency a part of its API? I don't know, check the documentation! Nothing that isn't a part of the public API can be relied upon, so it doesn't make any sense to version it in the first place. That's why the API is the SemVer rule number 1!
---
Phew! I hope you have a better idea of what software versioning is now, and understand that how a software package _uses_ its dependencies doesn't matter as long as its API doesn't change. After all, the API is all you ever care about as a user!

**Comment by elliottslaughter** [comment]
Summarizing what's been said so far:

1. Everyone agrees that exposing the version string (via `-v` or otherwise) is useful.

2. Three versioning schemes have been proposed: calver, semver, and explicit versioning. Calver is what Terra has done historically, but I think everyone agrees is suboptimal. There was a lot of back and forth on the thread about semver. However, I don't see discussion of alternatives, only a brief mention of explicit versioning.

My own take:

Terra is a volunteer project with no full-time staff. Whatever we do, first and foremost, we have to respect the amount of effort that Terra receives. Versioning requires effort. Thinking systematically about breaking changes requires effort[1]. Even just reading and responding to comment threads like this requires effort. For sure these are great things to have, but if they require more effort than is available to be spent on them, it's just not going to fly. (Case in point: see how long it took me to respond to this thread.)

We're not even in a position to have releases right now. We're closer than we used to be, with the automated CI, but still not all the way there. At a minimum, we'd want to figure out how to make the CI build the release artifacts, since there is no way I'm maintaining a build farm.

I'd like to get to a point where the project is healthy overall before we commit to a new version scheme. Top priorities in my mind are addressing the fragility in the build system, and doing something about the situation with LLVM. Unfortunately, I regularly run out of time even for responding to issues, so I'm not sure when any of it will happen. Until then, if we even have releases, I recommend sticking with calver since it's the status quo and punting on the long-term problem until the language is in a better place.

Y'all are welcome to show me that I'm wrong, but it'll have to involve some amount of stepping up to actually fix things, and not just arguing about what ought to be done (on some theoretical person's time).

[1]: Rust semver works because there is a [comprehensive document that describes exactly what is considered to be at each level of change](https://github.com/rust-lang/rfcs/blob/master/text/1105-api-evolution.md). And that's just for library APIs, not the language.

**Comment by elliottslaughter** [comment]
#314 makes a first pass at doing the version string part of this. It intentionally defers any decision about what versioning scheme to follow. But at least it's possible to query the version string now.

Please take a look and see if it will server your purposes or not.

**Comment by viluon** [comment]
Thanks for the very sane and reasonable replies, @elliottslaughter. Thanks even more for #314. While I understand your decision, I should also point out that the [semver major version zero](https://semver.org/#spec-item-4) exists specifically to address the in-dev state that Terra seems to currently be in.

**Comment by elliottslaughter** [comment]
So my personal opinion is that out of all the schemes discussed in this thread, semver is the most sane. It's not perfect, but of the issues listed in this thread it's not obvious to me at least that any other versioning scheme does any better. If there's anything to complain about, it might be that semver can be seen as over-promising what mere humans can be realistically expected to deliver, but I'm not sure how to fix that other than by setting reasonable expectations in the first place, and maybe not relying so much on automated version selection where it doesn't make sense (which Terra doesn't have any of anyhow).

Anyway, my objection is not so much to the version scheme, as just with the cognitive load associated with using it. Usually semver implies a stigma against making breaking changes. So now every change requires some amount of thought about whether it ought to be considered breaking or not.

For Terra, this gets particularly pernicious because there are a lot of ways the language could break. For example, adding a new LLVM version isn't breaking, because no one is forced to use it. However, removing support for an older LLVM version might be. But at some point, it seems like we'll have to---I already maintain patches to make LLVM 3.5 compile on modern machines. And Terra has code that goes back all the way to LLVM 3.2. At what point do those become old enough that we say we can remove them without considering it to be breaking?

You could increment the major version number on every release, (or the minor version in the 0.x series), but that doesn't seem very reasonable to me. First of all, Terra isn't changing that fast. Bumping a major version seems to communicate major changes, when there aren't any. Plus, I'm kind of pissed off at LLVM for refusing to retain backwards compatibility, and I don't want to do that to Terra's users (and myself, since I'm one of them).

The fact is, Terra *isn't* changing that quickly. We could almost just say that the next release is 1.0, and then there will be no breaking changes going forward. But I'm not sure I'm willing to promise no breaking changes, because the language hasn't necessarily seen the level of usage that would give us confidence that it does everything it needs to do right now. And yet, for the level of change it does see, bumping version numbers seems too much.

I'm willing to consider it, but we'll have to figure out what it means to be backwards compatible and how much effort will go into this.

**Comment by elliottslaughter** [comment]
I discussed this with @zdevito and he's ok with switching to semver with the next release as 1.0. So that's the plan, once our testing of the beta release is complete.

One question is whether we should put together a document that says what the stability guarantee means. E.g. I think we probably want to have cut-outs for the build system and certain dependencies, otherwise e.g. keeping up with LLVM versions won't be very reasonable. And even LuaJIT may become a maintenance burden at some point if it continues the way it's been going. I may write this up when I have the time.

I will close this issue for now since the version number patch has been accepted, but please feel free to comment if there is more to discuss.
